### ISSUE TITLE
Consider adding the reason for existence of both result and outcome

### ISSUE THREAD

**Issue by do-m-en** [ned14/outcome#80]
I haven't gone through the entire documentation for now but I'm fairly confident that this is either missing from the documentation or placed in a hard to find location...

From what I could gather from the tutorial the main reason for the existence of both result<T, EC> and outcome<T, EC, EP> is that with:

- result<T, EC>: the author of the function expects the user to either handle the error at the call site, determine that the error can be propagated upwards without the loss of information (example returning the result one level up will not confuse the user of that function which file failed to open as there is only one file being opened in the propagating function) or repackage the error to before returning the result/throwing an exception to something that is relevant to the callee (this is the usual case with low level errors such as opening files)

- outcome<T, EC, EP>: the author of the function expects the error to be specific enough to be relevant even if it is not handled directly at the call site (example of this would be that XML validation failed due to missing schema file xyz, due to error on line x, node referred to by the XPath is missing in the XML... where you in most cases don't need to add additional data to the exception to be understandable by the upper levels - for example by the roll-back-transaction-and-log-the-error-before-continuing-with-processing-of-the-next-package somewhere at the top of the callstack)

Since the distinction and the reason for the relevance of why both exist and where each should be used is important I'd add such an example towards the beginning of the documentation (and maybe a reference link to the section also at the beginning of the tutorial and the FAQ section). Maybe this should be added to the "Decision matrix" section (which should probably be renamed to "When to use result/outcome/exceptions" or something similar as "Decision matrix" was far from the first link that I've clicked on due to the current name...).

**Comment by ned14** [comment]
@akrzemi1 I don't want to comment on this until you comment on it - you had some difficulty originally distinguishing the need for both `result` and `outcome`. Can you perhaps respond to the above, and from that we'll try to devise a better, clearer wording of rationale for the documentation?

**Comment by akrzemi1** [comment]
I can see a question and a request. question: why two types and wen to use them. Request: put the response to the question in the visible place in the docs. Now, I am only replying to the question.

It is my understanding that the primary motivation for having two types is performance trade-offs. `result<T, EC>` gives you the guarantee that when `T` and `EC` are trivially-copyable (`memcpy` can be safely substituted for copy/move constructors/assignments) then `result<T, EC>` is also trivially-copyable. This guarantees fast returning by value, even across libraries. A user can often provide a trivially-copyable  `T` and `EC`; e.g., when `T` is an `int` and `EC` is an `std::error_code`, or an enum.

`outcome<T, EC, EP>`, on the other hand will not give you the trivially-copyable guarantee even if `T` and `EC` and `EP` are trivially-copyable, because it is designed and optimized for the case where `EP` is either `std::exception_ptr` or `boost::exception_ptr`.

It is also my understanding, that the information saying that "you can use `EC` for recoverable errors and `EP` for non-recoverable errors" or "expected and unexpected errors" is secondary; it is a convenience invented after the technical decisions had been made.

The advice for the user is, it *only you* are producing the error value, and it is your decision what value (of what type) it will be, and you can afford to produce the `EC` that is trivially-copyable, you should go with `result<T, EC>`. On the other hand, when your function calls other functions, and these other functions can either throw or return `outcome<T, EC, EP>`, then you actually have no choice but to go with `outcome<T, EC, EP>`.

Nial, is this correct?

**Comment by ned14** [comment]
Yes, that's correct. But ultimately the true rationale is "because you'll eventually find a need for a three-state object in any large codebase, and better that we supply one which works well than you going off and mashing something up which is inferior".

How do we succinctly, and without being patronising, summarise the rationale for `outcome` in the docs? Suggestions - or pull requests - are welcome. I'm pretty stumped, myself.

**Comment by akrzemi1** [comment]
What do you mean by a "three-state object"?
Either proper value or anticipated error or unanticipated error? Or something else?

**Comment by ned14** [comment]
It's more that eventually you'll find yourself typing `result<T, variant<EC, EP>>` in any large codebase and in our rationale for "Why `outcome`?", I think one needs to make crystal clear somehow why `outcome` is superior to `result<T, variant<EC, EP>>`. Does that make sense?

I'll put it another way, that question will have to be answered in the peer review in any case.

**Comment by akrzemi1** [comment]
To some extent it has been addressed [here](https://ned14.github.io/outcome/tutorial/outcome/). But maybe this needs to be more visible.

**Comment by akrzemi1** [comment]
I am trying to write up the rationale, but I need some help. In `outcome<T, EC, EP>` does it make any sense at all to substitute for `EP` anything but `std::exception_ptr` or `boost::exception_ptr`?

**Comment by ned14** [comment]
Yes, if combined with a custom NoValuePolicy. It can be any type you like doing anything you like. Obviously it's still observed via `.exception()`, so you probably ought to use something exception-y.

**Comment by akrzemi1** [comment]
Niall: Is the following correct?

## Why two types `result<>`  and `outcome<>` rather than just one?

The two types represent slightly different things. Using two types allows expressing programmer's intentions more directly.

Function signature `result<T> f() noexcept` means that any failures in `f` are exclusively reported by returning an error code. 

Similarly, signature `result<T> f()` means that in addition to the above, some failures in `f` are signalled by throwing exceptions. 

Now, signature `outcome<T> g() noexcept` means that in the lower layers of the implementation we have two kinds of libraries/modules: some report failures via error codes, and some by throwing exceptions. But at the level of our interface we do not want to throw exceptions; so we forward both error codes and exceptions unaltered. The callers can easily tell which type of failure reporting was chosen, and extract it appropriately.

This implies that `outcome<>` has different interface that allows you to ask whether we are storing an exception or an error code to rethrow the stored exception but to return by value the error code, etc.

There is one additional possibility. When you know there will be no error codes returned -- only exceptions thrown, but at some layer you want to temporarily change the stack unwinding into returning failures through return values, you can use signature

```
result<T, std::exception_ptr> f() noexcept;
```

**Comment by akrzemi1** [comment]
> I think the only missing bit really is that one can inject, unintrusively, rules for converting some `result` to some `outcome` and if we did not have `outcome`, we could not do this.

Now that you have included conversions between different `result`s inside `TRY` I do not think it is an argument any longer. I could always use `variant`:

```c++
out::result<int, std::error_code> f() { return std::errc::bad_address; }

out::result<int, std::variant<std::error_code, std::exception_ptr>> g()
{
    OUTCOME_TRY (i, f());
    return i;
}
```

> Also, a key thing Peter Dimov wanted was the ability to return BOTH an error code and exception pointer, and calling code can use either. outcome enables that.

Could you expand on this? When would this be useful?

**Comment by ned14** [comment]
> Now that you have included conversions between different results inside TRY I do not think it is an argument any longer. I could always use variant:

You could, though the default action for variant is probably UB (it might try to extract the error code and throw it as a system error, I'm not sure)

Remember Outcome's facilities here allow custom conversion, so a non 1-1 mapping. And without source code modification of third party code.

> Could you expand on this? When would this be useful?

Peter felt there was a use case for APIs which return both. Then low latency callers could use the error code, code which doesn't care could use the exception rethrow. It assumes that `std::make_exception_ptr` is fairly fast which wasn't the case during the Outcome v1 review, but I note all the major STLs fixed their implementation after the Outcome v2 review. So now it's true.

**Comment by akrzemi1** [comment]
> > Now that you have included conversions between different results inside TRY I do not think it is an argument any longer. I could always use variant:

> You could, though the default action for variant is probably UB (it might try to extract the error code and throw it as a system error, I'm not sure)
> 
> Remember Outcome's facilities here allow custom conversion, so a non 1-1 mapping. And without source code modification of third party code.

We are talking about this in the context of explaining "why two types rather than one". You say it is (also) because one can apply hooks for converting from one `result` to another `outcome`. Which sounds like, unless you have two `outcome` and `result` you cannot apply the conversion hooks. But don't you also provide hooks for converting from one `result` to another `result`?

**Comment by ned14** [comment]
The design plan, brought over from v1 Outcome, is that code generated without explicit C++ source behind it is always cheap as possible in terms of compile time and run time. Only when the programmers types an explicit something may we impose additional cost in extra compile time and/or run time. This is why Outcome goes to such measures to ensure that non-converting copies and moves of the same type are as cost free as possible.

Now, `outcome` almost always will appear in application code, and rarely in library code, because it is an information-preserving accumulator of whatever weird `result` configurations chosen by libraries. Or at least that's how it's ended up in my own code, generally where you're using `outcome`, you're not caring too much about performance - though you have that option to do so if you choose.

Does that help?

**Comment by akrzemi1** [comment]
Could you also give me a yes/no answer? Can `TRY` convert from `result<T, EC1>` to `result<T, EC2>` through the hook mechanism?

**Comment by ned14** [comment]
I believe not.

**Comment by akrzemi1** [comment]
I have added an entry to FAQ. See commit https://github.com/ned14/outcome/commit/c906272c635e7eea6d555fb375e978fb8b7c92e3. I have also updated the tutorial on `outcome<>` to reflect what we discussed here. I did not mention the performance considerations, as I do not fully understand your explanations. I gather that `result` is fast and `outcome` slower, but it does not fit as rationale for having two types.

@do-m-en: does the rationale in the FAQ make sense to you?
